### PR TITLE
feat: add db support for flatfee/usagebased charge intent overrides

### PR DESCRIPTION
**Pull request by turip**
## Summary

Adds type-owned charge intent overrides for flat-fee and usage-based charges.

Charge intents are the durable instructions used to produce charge state and invoice lines. Overrides are API-owned replacement intents for the same charge, so API edits can remain effective while subscription sync continues reconciling the base intent.

Overrides are full snapshots, not sparse patches. If no override row exists, the base intent is effective. If an override row exists, the override intent is effective.

## Rules

- Override presence is represented by a dedicated override row existing.
- Flat-fee and usage-based charges each own their override domain object and table.
- Credit-purchase charges do not participate in this override model.
- Creating an override creates the type-specific override row.
- Updating an override requires the override row to already exist.
- Deleting an override removes the override row and returns the charge to base-intent behavior.
- Override rows persist their own `invoice_at`; invoice-state calculations still use the base/original intent timing in this PR.

## Deletion Rules

Flat-fee and usage-based charges track deletion intent separately for the base intent and override intent:

- `Intent.IntentDeletedAt` marks the base/original intent as deleted.
- `IntentOverride.IntentDeletedAt` marks the override intent as deleted.
- If no override row exists, adapters derive the charge's effective `DeletedAt` from `Intent.IntentDeletedAt`.
- If an override row exists, adapters derive the charge's effective `DeletedAt` from `IntentOverride.IntentDeletedAt`.

This allows subscription/system state to delete the base intent while an API override remains effective, and allows API override deletion without mutating the base intent.

## Summary by CodeRabbit

* **New Features**
  * Added charge intent override support for flat-fee and usage-based charges (create/delete), with correct effective deletion timestamp behavior.
  * Charge lists/search now include the effective base intent-deleted timestamp for supported charge types.
* **Bug Fixes**
  * Charge update/delete logic now resolves and persists `DeletedAt` and intent-deleted timestamps consistently (UTC-normalized), including when overrides exist.
  * Reads handle override associations correctly and ignore stale override fields; credit-purchase remains excluded from override semantics.
  * Credit-then-invoice extend/shrink now reconciles invoicing state more reliably.
* **Documentation**
  * Added “intent deletion rules” documentation for effective intent-deleted timestamp resolution.
* **Tests**
  * Added end-to-end coverage for the full intent-override lifecycle and key edge cases for both flat-fee and usage-based charges.

### CODE REVIEW

File: openmeter/billing/charges/flatfee/adapter/intentoverride.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+func (a *adapter) updateIntentOverride(ctx context.Context, chargeID meta.ChargeID, override *flatfee.IntentOverride) (*entdb.ChargeFlatFeeOverride, error) {
+	if err := chargeID.Validate(); err != nil {
+		return nil, fmt.Errorf("charge id: %w", err)
+	}
+
+	normalized := override.Normalized()
+	if err := normalized.Validate(); err != nil {
+		return nil, fmt.Errorf("validating intent override: %w", err)
+	}
+
```
[the remark this answers is not in the document]
Comment (turip) [reply openmeter/billing/charges/flatfee/adapter/intentoverride.go:172]: Intentionally leaving this as the raw Ent/DB constraint error. Duplicate override creation is caller misuse rather than a domain branch we want callers to depend on, and wrapping it would add an adapter-specific error surface without value here. The test only asserts that the duplicate create fails.

File: .agents/skills/ent/SKILL.md [hunk 2]
```diff
@@ -35,13 +35,20 @@ After any schema change, regenerate with `make generate` before running tests.
...
+- **Generated `SetOrClear<Field>` helpers**: prefer these helpers for nullable/optional update fields when they have a straightforward signature. For awkward generated signatures such as double-pointer JSON fields, prefer the explicit pattern used for fields like normalized metadata: `if value != nil { update = update.Set<Field>(value) } else { update = update.Clear<Field>() }`. This avoids passing the address of a nil pointer, which can make Ent treat the field as set with a nil value and then panic or fail in generated validators.
 
 ## Regeneration
 
+Depending on the change, the generators needs to be re-run. For schema changes it's usually enough to regenerate the ent schema directly (use direnv as needed):
```
[the remark this answers is not in the document]
Comment (turip) [reply .agents/skills/ent/SKILL.md]: Fixed in 2297ae4cf. The grammar in the regeneration section now uses "generators need" and the surrounding wording was tightened.

File: .agents/skills/ent/SKILL.md [hunk 3]
```diff
@@ -35,13 +35,20 @@ After any schema change, regenerate with `make generate` before running tests.
...
+
+```bash
+# Regenerate Ent only after schema changes
+go generate ./openmeter/ent/...
+```
+
```
[the remark this answers is not in the document]
Comment (turip) [reply .agents/skills/ent/SKILL.md]: Fixed in 2297ae4cf. The skill now makes `make generate` the primary schema-change command and keeps the narrower Ent-only `go generate ./openmeter/ent/...` path only as a local iteration shortcut.

File: openmeter/billing/charges/flatfee/adapter/intentoverride.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+	dbOverride, err := a.db.ChargeFlatFeeOverride.Query().
+		Where(dbchargeflatfeeoverride.NamespaceEQ(chargeID.Namespace)).
+		Where(dbchargeflatfeeoverride.ChargeIDEQ(chargeID.ID)).
+		Only(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("querying updated intent override: %w", err)
+	}
+
+	return dbOverride, nil
```
[the remark this answers is not in the document]
Comment (turip) [reply openmeter/billing/charges/flatfee/adapter/intentoverride.go:220]: Intentionally not changing this. These helpers are private and are only called from transaction-aware public adapter methods in this package. Duplicating `TransactingRepo` inside the private helpers would add noisy layering and does not match the adapter convention we use here.

File: tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql [hunk 5]
```diff
@@ -0,0 +1,72 @@
+-- modify "charge_flat_fees" table
+ALTER TABLE "charge_flat_fees" DROP COLUMN "override_kind", ADD COLUMN "override_present" boolean NOT NULL DEFAULT false, ADD COLUMN "override_intent_deleted_at" timestamptz NULL, ADD COLUMN "intent_deleted_at" timestamptz NULL;
+-- modify "charge_usage_based" table
+ALTER TABLE "charge_usage_based" DROP COLUMN "override_kind", ADD COLUMN "override_present" boolean NOT NULL DEFAULT false, ADD COLUMN "override_intent_deleted_at" timestamptz NULL, ADD COLUMN "intent_deleted_at" timestamptz NULL;
```
[the remark this answers is not in the document]
Comment (turip) [reply tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql:4]: Intentionally not adding a guard/backfill or `RAISE EXCEPTION` migration here. The old override columns are unused for this implementation path and there is no override data to preserve, so dropping `override_kind` directly keeps this migration simpler and avoids unnecessary SQL churn.

File: tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
```
[the remark this answers is not in the document]
Comment (turip) [reply tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql:4]: Intentionally not adding a guard/backfill for this migration. The old inline override marker is unused in this implementation path and there is no live override data to preserve, so keeping the migration as direct DDL avoids unnecessary SQL churn. This mirrors the CodeRabbit thread above that we already discussed and resolved.

File: openmeter/billing/charges/flatfee/adapter/charge.go [hunk 7]
```diff
@@ -70,11 +72,22 @@ func (a *adapter) UpdateCharge(ctx context.Context, charge flatfee.ChargeBase) (
 			return flatfee.ChargeBase{}, err
```
[the remark this answers is not in the document]
Comment (turip) [reply openmeter/billing/charges/flatfee/adapter/charge.go:72]: Intentionally not adding a defensive DB lookup in `UpdateCharge`. `UpdateCharge` updates the aggregate it is given; if an override exists, callers must pass the aggregate with its current override state loaded. Adding a read here would make every update pay for a caller-contract bug and would hide the real issue in the path that drops override state before calling the adapter.

File: tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+-- create index "chargeflatfeeoverride_namespace" to table: "charge_flat_fee_overrides"
+CREATE INDEX "chargeflatfeeoverride_namespace" ON "charge_flat_fee_overrides" ("namespace");
+-- create index "chargeflatfeeoverride_id" to table: "charge_flat_fee_overrides"
+CREATE UNIQUE INDEX "chargeflatfeeoverride_id" ON "charge_flat_fee_overrides" ("id");
+-- create index "charge_flat_fee_overrides_charge_id_key" to table: "charge_flat_fee_overrides"
+CREATE UNIQUE INDEX "charge_flat_fee_overrides_charge_id_key" ON "charge_flat_fee_overrides" ("charge_id");
+-- create index "chargeflatfeeoverrides_tax_code_id" to table: "charge_flat_fee_overrides"
+CREATE INDEX "chargeflatfeeoverrides_tax_code_id" ON "charge_flat_fee_overrides" ("tax_code_id");
+-- create index "chargeflatfeeoverride_namespace_charge_id" to table: "charge_flat_fee_overrides"
+CREATE UNIQUE INDEX "chargeflatfeeoverride_namespace_charge_id" ON "charge_flat_fee_overrides" ("namespace", "charge_id");
```
Comment (borosr) [inline tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql:41]: Because this composite index contains the namespace and the charge id, why do we need the previous indexes on each columns separately?

File: tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  CONSTRAINT "charge_flat_fee_overrides_tax_codes_charge_flat_fee_overrides" FOREIGN KEY ("tax_code_id") REFERENCES "tax_codes" ("id") ON UPDATE NO ACTION ON DELETE SET NULL
+);
+-- create index "chargeflatfeeoverride_namespace" to table: "charge_flat_fee_overrides"
+CREATE INDEX "chargeflatfeeoverride_namespace" ON "charge_flat_fee_overrides" ("namespace");
+-- create index "chargeflatfeeoverride_id" to table: "charge_flat_fee_overrides"
+CREATE UNIQUE INDEX "chargeflatfeeoverride_id" ON "charge_flat_fee_overrides" ("id");
+-- create index "charge_flat_fee_overrides_charge_id_key" to table: "charge_flat_fee_overrides"
+CREATE UNIQUE INDEX "charge_flat_fee_overrides_charge_id_key" ON "charge_flat_fee_overrides" ("charge_id");
+-- create index "chargeflatfeeoverrides_tax_code_id" to table: "charge_flat_fee_overrides"
+CREATE INDEX "chargeflatfeeoverrides_tax_code_id" ON "charge_flat_fee_overrides" ("tax_code_id");
```
Comment (borosr) [inline tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql:39]: Is the tax code id going to be queried individually or does it make sense to make this index a composite index like this: `("namespace", "tax_code_id")`?

**Review by borosr — APPROVED** [review]
Other then the two question I sent the code looks good

File: tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql (hunk 8 above)
Comment (turip) [reply tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql:41]: We have the specific indexes to satisfy any FK constraints. But thrutfully we are adding more indexes than we should. This is a pseudo decision that we take, that we usually add both the composite and the normal index.

(Ent also queries both by ID and id + namespace)

File: tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql (hunk 9 above)
Comment (turip) [reply tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql:39]: It's only there because of FK constraints.

File: tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql (hunk 8 above)
Comment (borosr) [reply tools/migrate/migrations/20260624043300_charge_intent_override_ddl.up.sql:41]: Thanks for the clarification, that makes sense
